cou/utils/juju_utils.py

```diff
@@ -396,3 +397,101 @@ async def async_upgrade_charm(
         revision=revision,
         switch=switch,
     )
+
+
+# pylint: disable=too-few-public-methods
+class JujuWaiter:
+    """Enhanced version of wait_for_idle.
+
+    Usage:
+        JujuWaiter(model).wait(120)
+    """
+
+    DEFAULT_TIMEOUT: int = 3600
+    MODEL_IDLE_PERIOD: int = 30
+    JUJU_IDLE_CHECK_PERIOD: int = 40
+
+    class TimeoutException(Exception):
+        """Own timeout exception."""
+
+    def __init__(self, model: Model):
+        """Initialize.
+
+        :param model: model to wait
+        :type model: Model
+        """
+        self.model = model
+        self.model_name = self.model.info.name
+        self.timeout = timedelta(seconds=JujuWaiter.DEFAULT_TIMEOUT)
+        self.start_time = datetime.now()
+        self.log = logging.getLogger(self.__class__.__name__)
+
+    async def wait(self, timeout_seconds: Optional[int] = None) -> None:
+        """Wait model to stabilize.
+
+        :param timeout_seconds: wait model till timeout_seconds. If passed raise TimeoutException
+        :type timeout_seconds: Optional[int]
+        :return: None
+        :raises TimeoutException: if timeout_seconds passed.
+        :raises JujuMachineError: if it has a machine exception
+        :raises JujuAgentError: if it has an agent exception
+        :raises JujuUnitError: if it has unit exception
+        :raises JujuAppError: if it has application exception
+        """
+        self.start_time = datetime.now()
+        if timeout_seconds:
+            self.timeout = timedelta(seconds=timeout_seconds)
+        self.log.debug("Waiting to stabilize in %s seconds", self.timeout)
+
+        while True:
+            await self._ensure_model_connected()
+            try:
+                await self.model.wait_for_idle(
+                    idle_period=JujuWaiter.MODEL_IDLE_PERIOD,
+                    timeout=JujuWaiter.JUJU_IDLE_CHECK_PERIOD,
+                )
+                self.log.debug(
+                    "Model %s is idle for %s seconds.",
+                    self.model.info.name,
+                    JujuWaiter.MODEL_IDLE_PERIOD,
+                )
+                return
+            except (
+                JujuMachineError,
+                JujuAgentError,
+                JujuUnitError,
+                JujuAppError,
+            ) as ex:
+                raise ex
+            except JujuWaiter.TimeoutException as ex:
+                self.log.debug(
+                    "Unable to stabilize in %s seconds",
+                    datetime.now() - self.start_time,
+                )
+                raise ex
```

Comment:
When this could happen? Libjuju will never raise your own exception.

Reply:
True. This should be under `_check_time()`

Reply:
I do not see purpose of this, since we have `_check_time` function, which is providing some logs and can provide this also, instead of catching it and raising it again.

Reply:
it is raised by `_check_time`

Reply:
But `_check_time ` is outside of try. And as I mention there is not reason to catching it here a provide log like this. This log message belong to the function itself.

Reply:
removed.